auto_submit/lib/requests/event_processor.dart

```diff
@@ -0,0 +1,159 @@
+// Copyright 2023 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'package:auto_submit/request_handling/pubsub.dart';
+import 'package:github/github.dart';
+import 'package:shelf/shelf.dart';
+
+import '../model/merge_comment_message.dart';
+import '../service/config.dart';
+import '../service/log.dart';
+
+final String nonSuccessResponse = jsonEncode(<String, String>{});
+
+abstract class EventProcessor {
+  factory EventProcessor(
+    String eventType,
+    Config config,
+    PubSub pubSub,
+  ) {
+    switch (eventType) {
+      case 'pull_request':
+        return PullRequestProcessor(config, pubSub);
+      case 'issue_comment':
+        return IssueCommentProcessor(config, pubSub);
+      default:
+        return NoOpRequestProcessor();
+    }
+  }
+
+  Future<Response> processEvent(List<int> requestBytes);
+}
+
+class NoOpRequestProcessor implements EventProcessor {
+  @override
+  Future<Response> processEvent(List<int> requestBytes) async {
+    return Response.ok(nonSuccessResponse);
+  }
+}
+
+class PullRequestProcessor implements EventProcessor {
+  PullRequestProcessor(this.config, this.pubSub);
+
+  final Config config;
+  final PubSub pubSub;
+
+  @override
+  Future<Response> processEvent(List<int> requestBytes) async {
+    bool hasAutosubmit = false;
+    bool hasRevertLabel = false;
+    final String rawBody = utf8.decode(requestBytes);
+    final Map<String, dynamic> body = json.decode(rawBody) as Map<String, dynamic>;
+
+    if (!body.containsKey('pull_request') || !((body['pull_request'] as Map<String, dynamic>).containsKey('labels'))) {
+      return Response.ok(nonSuccessResponse);
+    }
+
+    final PullRequest pullRequest = PullRequest.fromJson(body['pull_request'] as Map<String, dynamic>);
+
+    hasAutosubmit = pullRequest.labels!.any((label) => label.name == Config.kAutosubmitLabel);
+    hasRevertLabel = pullRequest.labels!.any((label) => label.name == Config.kRevertLabel);
+
+    if (hasAutosubmit || hasRevertLabel) {
+      log.info('Found pull request with auto submit and/or revert label.');
+      await pubSub.publish(
+        'auto-submit-queue',
+        pullRequest,
+      );
+      return Response.ok(rawBody);
+    }
+
+    return Response.ok(nonSuccessResponse);
+  }
+}
+
+class IssueCommentProcessor implements EventProcessor {
+  IssueCommentProcessor(this.config, this.pubSub);
+
+  final Config config;
+  final PubSub pubSub;
+
+  /// Process a github issue comment that was passed to this webhook.
+  ///
+  /// In order for the comment to be processed it must be a newly created
+  /// comment, and be a comment left on a pull request. The author of the
+  /// comment must also be a MEMBER or OWNER of the repository.
+  @override
+  Future<Response> processEvent(List<int> requestBytes) async {
+    log.info('Processing event...');
```

Comment:
There is common logic in the processEvent implementations, should this logic go to the abstract class?

Reply:
There are only really two lines in common. If I put them in a common class we need access to both the raw body and the resulting map. Right now that would be more lines of code than the duplicated code combined.

Reply:
I don't believe Dart allows impls in an abstract class does it?